### PR TITLE
ARROW-16725: [C++] Fix compilation warnings in release mode

### PR DESCRIPTION
**Pull request by pitrou**
As seen on e.g. MinGW CI builds.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/vector_sort.cc [hunk 1]
```diff
@@ -2006,10 +2006,11 @@ class ArrayRanker : public TypeVisitor {
       }
 
       case RankOptions::Min: {
-        T curr_value, prev_value;
+        T curr_value, prev_value{};
+        rank = 0;
 
         if (null_placement_ == NullPlacement::AtStart) {
-          rank = 1;
+          rank++;
```
Comment (westonpace) [inline cpp/src/arrow/compute/kernels/vector_sort.cc:2013]: This seems harmless but why this change?

File: cpp/src/arrow/compute/exec/ir_consumer.cc [hunk 2]
```diff
@@ -625,7 +625,7 @@ Result<Declaration> Convert(const ir::Relation& rel) {
             break;
         }
 
-        NullPlacement key_null_placement;
+        NullPlacement key_null_placement{};
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/ir_consumer.cc:628]: Perhaps we could add a default case to the switch statement instead?  Although I think this code is going away so probably not worth it to nitpick.

**Review by westonpace — APPROVED** [review]

**Comment by kou** [comment]
It seems there are more warnings. e.g.:

https://github.com/apache/arrow/runs/6707180740?check_suite_focus=true#step:7:1074

```text
In file included from D:/a/arrow/arrow/build/cpp/src/arrow/compute/kernels/CMakeFiles/arrow-compute-scalar-test.dir/Unity/unity_0_cxx.cxx:3:
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc: In instantiation of 'void arrow::compute::TestUnaryArithmeticFloating_Sqrt_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = arrow::DoubleType]':
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:3338:1:   required from here
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:3342:8: warning: unused variable 'max_val' [-Wunused-variable]
 3342 |   auto max_val = std::numeric_limits<CType>::max();
      |        ^~~~~~~
```

Do you want to suppress them in this pull request?

**Comment by kou** [comment]
It seems there are more warnings. e.g.:

https://github.com/apache/arrow/runs/6707180740?check_suite_focus=true#step:7:1074

```text
In file included from D:/a/arrow/arrow/build/cpp/src/arrow/compute/kernels/CMakeFiles/arrow-compute-scalar-test.dir/Unity/unity_0_cxx.cxx:3:
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc: In instantiation of 'void arrow::compute::TestUnaryArithmeticFloating_Sqrt_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = arrow::DoubleType]':
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:3338:1:   required from here
D:/a/arrow/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:3342:8: warning: unused variable 'max_val' [-Wunused-variable]
 3342 |   auto max_val = std::numeric_limits<CType>::max();
      |        ^~~~~~~
```

Do you want to suppress them in this pull request?

File: cpp/src/arrow/compute/kernels/vector_sort.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/vector_sort.cc:2013]: Just for consistency for other similar code blocks above.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp
